src/Utilities/Compiler/Extensions/ITypeSymbolExtensions.cs

```diff
@@ -135,18 +135,71 @@ public static bool ImplementsIDisposable(this ITypeSymbol type, INamedTypeSymbol
         public static bool IsDisposable(this ITypeSymbol type, INamedTypeSymbol iDisposable)
             => type.IsReferenceType && (Equals(type, iDisposable) || type.ImplementsIDisposable(iDisposable));
 
-        public static IEnumerable<AttributeData> GetApplicableAttributes(this INamedTypeSymbol type)
+        public static IEnumerable<AttributeData> GetApplicableAttributes(this INamedTypeSymbol type, INamedTypeSymbol attributeUsageAttribute)
         {
             var attributes = new List<AttributeData>();
+            var includeInherited = true;
 
             while (type != null)
             {
-                attributes.AddRange(type.GetAttributes());
+                var current = type.GetAttributes();
+                if (includeInherited || attributeUsageAttribute is null)
+                {
+                    attributes.AddRange(current);
+                }
+                else
+                {
+                    foreach (var attribute in current)
+                    {
+                        if (!IsInheritedAttribute(attribute, attributeUsageAttribute))
+                        {
+                            continue;
+                        }
+
+                        attributes.Add(attribute);
+                    }
+                }
 
                 type = type.BaseType;
+                includeInherited = false;
```

Comment:
The flag name is confusing, as it seems to be false for all the base types, but then we force include only inherited attributes for base types in the foreach loop above. Either its values should be inverse or the flag name should be changed.

Reply:
Inverted the behavior and renamed to `onlyIncludeInherited`